symengine/numerics/optimizer.h

```diff
@@ -0,0 +1,1444 @@
+     * counterpart of all dependent functions. This function fetches those
+     * cached values.
+     *
+     * These values were computed by substituting a @p substitution_values map
+     * during substitute() call.
+     */
+    const std::vector<ReturnType> &get_evaluated_dependent_expressions() const;
+
+    /**
+     * Returns the result of a value substitution into the optimized
+     * counterpart of @p func. This function fetches that cached value.
+     *
+     * This value was computed by substituting a @p substitution_values map
+     * during substitute() call.
+     */
+    ReturnType evaluate(const Expression &func) const;
+
+    //@}
+
+private:
+    /**
+     * The optimization methods that is to be employed.
+     */
+    enum OptimizerType method;
+
+    /**
+     * The optimization flags that indicate which expression manipulation
+     * mechanisms are to be employed.
+     */
+    enum OptimizationFlags flags;
+
+    /**
+     * A map that represents the symbols that form the set of independent
+     * variables upon which optimized symbolic expressions are to be based.
+     *
+     * @note As the ordering of the input symbols is fixed at the time at
+     * which optimization is performed, we store all of the entries in
+     * a map to ensure that both we and the user never mistakenly
+     * swap the order of two or more symbols during evaluation.
+     */
+    SymEngine::map_basic_basic independent_symbols;
+
+    /**
+     * A set of symbolic expressions to be optimized. It is required that
+     * the symbols on which these dependent functions be based are are
+     * registered in the @p independent_symbols map.
+     */
+    SymEngine::vec_basic dependent_expressions;
+
+    /**
+     * The output of substituting symbolic values with floating point
+     * values through the use of the @p optimizer.
+     *
+     * It is necessary to use this intermediate storage mechanism
+     * to store the result of a substitution sweep as some optimizers
+     * work on all symbolic expressions in a single batch. In this way
+     * they can employ methods such as common subexpression elimination
+     * to minimise the number of terms evaluated across all symbolic
+     * functions.
+     *
+     * This variable is marked as mutable. This facilitates the substitution
+     * functionality being used in logically constant "get_*" functions.
+     */
+    mutable std::vector<ReturnType> evaluated_dependent_expressions;
+
+    /**
+     * A map type used to indicate which dependent variable is associated
+     * with which entry in the output vector.
+     *
+     * We use a custom comparator here because otherwise we can't use
+     * std::map::find; it is sensitive to the some data from the
+     * underlying SymEngine::Basic other than the value that it represents.
+     */
+    using map_dependent_expression_to_vector_entry_t
+        = std::map<const SymEngine::Basic *const, std::size_t,
```

Comment:
Instead of storing raw pointers, use RCP here for reference counting

Reply:
I believe that I tried this at first, but it didn't work correctly in this context:
```c++
const auto &it = map_dep_ptr_basic_vec_entry.find(&*(func.get_basic()));
```
As best as I can recall, if one uses `RCP<Basic>` then `find` tries to match the `RCP` and not the underlying `Basic`.  However, I'll try again to switch this over to using `RCP<Basic>` and see if I can make it work.

Reply:
I think that the missing element when I first implemented this was to use `RCPBasicKeyLess` as the comparator. I've made the change to `
std::map<RCP<const Basic>, std::size_t, RCPBasicKeyLess>` and I think that this now works as expected. Thanks for the suggestion!